charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,249 @@
+                    "command": exec_command,
+                    "startup": "enabled",
+                    "environment": self._env_vars,
+                    "on-check-failure": {"tensorboards-web-app-up": "restart"},
+                }
+            },
+            "checks": {
+                "tensorboards-web-app-up": {
+                    "override": "replace",
+                    "period": "30s",
+                    "http": {"url": f"http://localhost:{PORT}"},
                 },
-                "containers": [
-                    {
-                        "name": "tensorboards-web-app",
-                        "imageDetails": image_details,
-                        "envConfig": {
-                            "USERID_HEADER": "kubeflow-userid",
-                            "USERID_PREFIX": "",
-                            "APP_PREFIX": "/tensorboards",
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
 
-        self.model.unit.status = ActiveStatus()
+        return Layer(layer_config)
 
-    def _configure_mesh(self, interfaces):
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as error:
+            self._log_and_set_status(error.status)
+        return
+
+    def _on_remove(self, _) -> None:
+        """Remove all resources."""
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        try:
+            self.k8s_resource_handler.delete()
+        except ApiError as error:
+            # Do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {error}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from error
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            update_layer(
+                self._container_name,
+                self._container,
+                self._tensorboards_web_app_layer,
+                self.logger,
+            )
+            interfaces = self._get_interfaces()
+            self._configure_mesh(interfaces)
+            self.unit.status = ActiveStatus()
+
+        except ErrorWithStatus as error:
+            self._log_and_set_status(error.status)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _deploy_k8s_resources(self) -> None:
```

Comment:
Just for correctness:

```suggestion
    def _apply_k8s_resources(self) -> None:
        """Applies K8S resources."""
```

This method is called in different places, which means that not all the times it's deploying k8s resources, but just patching (if any change).